crates/sui-transactional-test-runner/src/test_adapter.rs

```diff
@@ -584,11 +588,103 @@ impl<'a> MoveTestAdapter<'a> for SuiTestAdapter<'a> {
+                            let named_addr_opt = named_addr_opt.map(|n| n.value);
+                            let module = annot_module.named_module.module;
+                            (named_addr_opt, module)
+                        }
+                        AnnotatedCompiledUnit::Script(_) => panic!(
+                            "Expected a module text block, not a script, \
+                                    following 'publish' starting on lines {}-{}",
+                            start_line, command_lines_stop
+                        ),
+                    })
+                    .collect();
+
+                let output =
+                    self.upgrade_package(package, modules, upgrade_capability, gas_budget)?;
+                Ok(merge_output(warnings_opt, output))
+            }
+        }
+    }
+}
+
+fn merge_output(left: Option<String>, right: Option<String>) -> Option<String> {
+    match (left, right) {
+        (None, right) => right,
+        (left, None) => left,
+        (Some(mut left), Some(right)) => {
+            left.push_str(&right);
+            Some(left)
         }
     }
 }
 
 impl<'a> SuiTestAdapter<'a> {
+    fn upgrade_package(
+        &mut self,
+        current_package_object_id: Option<String>,
+        modules: Vec<(Option<Symbol>, CompiledModule)>,
+        upgrade_capability: Option<u64>,
+        gas_budget: Option<u64>,
+    ) -> anyhow::Result<Option<String>> {
+        let current_package_object_id = current_package_object_id
+            .map_or(ObjectID::from_address(AccountAddress::ZERO), |id| {
+                ObjectID::from_str(&id).unwrap()
+            });
+        let upgrade_capability = upgrade_capability.unwrap_or_default();
+        let all_bytes = modules
+            .iter()
+            .map(|(_, module)| {
+                let mut module_bytes = vec![];
+                module.serialize(&mut module_bytes)?;
+                Ok(module_bytes)
+            })
+            .collect::<anyhow::Result<_>>()?;
+        let gas_budget = gas_budget.unwrap_or(DEFAULT_GAS_BUDGET);
+        let dependencies = system_package_ids();
+        let data = |sender, gas| {
+            let mut builder = ProgrammableTransactionBuilder::new();
+            builder.upgrade(
+                current_package_object_id,
+                Argument::Input(upgrade_capability as u16),
+                dependencies,
+                all_bytes,
+            );
+            let pt = builder.finish();
+            TransactionData::new_programmable_with_dummy_gas_price(
+                sender,
+                vec![gas],
+                pt,
+                gas_budget,
+            )
+        };
+
+        let transaction = self.sign_txn(None, data);
+        let summary = self.execute_txn(transaction, gas_budget)?;
+        let view_events = false;
+        let output =
+            self.object_summary_output(&summary, view_events, /* view_gas_used */ false);
+        Ok(output)
```

Comment:
Note: this only returns txn output right now, we could return more things for upgraded outputs, but trying to keep changes small / self-contained.